clarity/evaluator/haspi/ip.py

```diff
@@ -4,7 +4,7 @@
 import numpy as np
 
 
-def get_neural_net() -> Tuple[Dict, List[np.ndarray], List[np.ndarray], float]:
+def get_neural_net() -> tuple[dict, list[np.ndarray], list[np.ndarray], float]:
```

Comment:
why change from `Tuple` to `tuple`, `Dict` to `dict` and so on?
Is it related to `pyupgrade`?

We may need to do this change throughout the whole repository. Including recipes.
Do we need to remove the importing of `typing` and add the `from __future__ import annotations`?

Reply:
Tuple, Dict etc were a hack for early versions of Python when type hints were first being  introduced. 

In modern python versions types are now built into the language as tuple, dict etc. So you don't need to import from Typing anymore.  And you can still use this in older versions of python as long as you include "from __future__ import annotations" at the start of the code.

Note, the modern type hints also allow things line "float | np.ndarray" rather than the clumpy "Union[float, np.ndarray]" and for "Optional" you can use e.g. "int | None" -- it's all much cleaner.

It's related to pyupgrade in the sense that if you do write hints in the obsolete style then pyupgrade will automatically fix them for you.

Yes, it has been changed throughout the repository and recipes.  pyupgrade has done that for us. :-)

(Note, pyupgrade will make it easy for us to keep the repo up to date. We just need to bump up the version number on its command line as we drop support for older (unsafe versions of Python) -- e.g. security fixes are no longer being made for 3.6 and will be dropped for 3.7 in a couple of months. i.e. we will shortly remove 3.7 support and hence be able to use features from 3.8 onwards)

Reply:
Thank you so much for the comprehensive message. Very clear.

It is great to know that  `pyupgrade` did all this tedious work for us.